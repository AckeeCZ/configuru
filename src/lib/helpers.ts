import * as jsonParser from 'jsonc-parser';

export const anonymize = (val: any) => {
    const str = String(val);
    const show = Math.min(str.length / 6, 10);
    return [str.slice(0, show), '***', str.slice(str.length + 1 - show, str.length)].join('');
};

export const parseBool = (x: any) => (x === 'false' || x === '0' ? false : Boolean(x));
export const identity = <T>(x: T) => x;

export const isObject = (x: any) => typeof x === 'object' && Object.prototype.toString.call(x) === '[object Object]';

export const JSONC = {
    parse: (text: string) => {
        const errors: jsonParser.ParseError[] = [];
        const parsed = jsonParser.parse(text, errors);
        if (errors.length) {
            throw new SyntaxError(`${jsonParser.printParseErrorCode(errors[0].error)} at position ${errors[0].offset}`);
        }
        return parsed;
    },
};
