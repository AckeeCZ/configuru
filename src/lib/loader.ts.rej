diff a/src/lib/loader.ts b/src/lib/loader.ts	(rejected hunks)
@@ -38,2 +36,2 @@ export const createAtomLoaderFactory = (storage: Record<any, any>) => {
-            value: missing ? null : transform(value) as any,
-        }
+            value: missing ? null : (transform(value) as any),
+        };
