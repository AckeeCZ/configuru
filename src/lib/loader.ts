import { createConfigStorage } from './storage';

// helpers
const parseBool = (x: any) => (x === 'false' || x === '0' ? false : Boolean(x));
const identity = <T>(x: T) => x;

export interface ConfigLoaderOptions {
    defaultConfigPath?: string;
    userConfigPath?: string;
    envMode?: boolean;
}

const defaultOpts = {
    defaultConfigPath: '.env.json',
    userConfigPath: process.env.CFG_JSON_PATH,
    envMode: true,
};

export const createLoaders = (opts: ConfigLoaderOptions = defaultOpts) => {
    const configStorage = createConfigStorage(opts);
    const load = <T, N extends boolean, R = N extends false ? T : T | null>(
        transform: (x: any) => T,
        anonymize: (x: any) => any,
        hidden: boolean,
        nullable: boolean
    ) => (key: string): R => {
        const value = configStorage[key];
        if (nullable && value === null) {
            return value;
        }
        if (value === undefined) {
            throw new Error(`Missing required value ${key}`);
        }
        if (hidden) {
            return anonymize(value);
        }
        return (transform(value) as any) as R;
    };
    const atomLoader = <T>(transform: (x: any) => T, anonymize: (x: any) => any = identity) =>
        Object.assign(load<T, false>(transform, anonymize, false, false), {
            hidden: Object.assign(load<T, false>(transform, anonymize, true, false), {
                nullable: load<T, true>(transform, anonymize, true, true),
            }),
            nullable: Object.assign(load<T, true>(transform, anonymize, false, true), {
                hidden: load<T, true>(transform, anonymize, true, true),
            }),
        });
    const anonymize = (val: any) => {
        const str = String(val);
        const show = str.length / 6;
        return [str.slice(0, show), '***', str.slice(str.length + 1 - show, str.length)].join('');
    };
    return {
        configLoader: {
            number: atomLoader(Number),
            string: atomLoader(String),
            bool: atomLoader(parseBool),
            json: atomLoader(JSON.parse),
        },
        anonymizedConfigLoader: {
            number: atomLoader(Number, anonymize),
            string: atomLoader(String, anonymize),
            bool: atomLoader(parseBool, anonymize),
            json: atomLoader(JSON.parse, anonymize),
        },
    };
};

export type ConfigLoader = ReturnType<typeof createLoaders>['configLoader'];
