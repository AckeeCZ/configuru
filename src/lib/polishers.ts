import { anonymize, isObject } from './helpers';
import { LoadedValue } from './loader';

type LoadedToValue<X> = X extends LoadedValue<any, any> ? X['value'] : X;

type Values<C> = C extends Function
    ? C
    : C extends object
    ? { [K in keyof C]: C[K] extends LoadedValue<any, any> ? LoadedToValue<C[K]> : Values<C[K]> }
    : C;

type AnonymousValues<C> = C extends Function
    ? C
    : C extends object
    ? { [K in keyof C]: C[K] extends LoadedValue<any, any> ? string : AnonymousValues<C[K]> }
    : C;

const isLoadedValue = (x: any) => Object.keys(x || {}).includes('__CONFIGURU_LEAF');

const mapConfig = (fn: (v: LoadedValue<any, any>) => any) => (val: any): any => {
    if (isLoadedValue(val)) {
        return fn(val);
    }
    if (Array.isArray(val)) {
        return val.map(mapConfig(fn));
    }
    if (isObject(val)) {
        return Object.keys(val).reduce((res: any, key) => {
            res[key] = mapConfig(fn)(val[key]);
            return res;
        }, {});
    }
    return val;
};

export const values = mapConfig(x => x.value) as <T extends Record<any, any>>(config: T) => Values<T>;
export const safeValues = mapConfig(x => (x.hidden ? anonymize(x.rawValue) : x.value)) as <T extends Record<any, any>>(
    config: T
) => AnonymousValues<T>;
