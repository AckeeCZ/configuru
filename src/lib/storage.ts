import { readFileSync, statSync } from 'fs'
import { format, join, parse, resolve } from 'path'
import { JSONC } from './helpers'
import { ConfigLoaderOptions } from './loader'

const fileExistsSync = (path: string) => {
  try {
    return statSync(path).isFile()
  } catch (_error) {
    return false
  }
}
const fromPairs = (pairs: Array<[keyof any, any]>) =>
  Object.assign({}, ...Array.from(pairs, ([k, v]) => ({ [k]: v })))
const uniq = <T>(xs: T[]) => Array.from(new Set(xs))
const { keys } = Object

const loadFile = (filePath?: string) => {
  if (!filePath) return {}
  const { dir, name } = parse(filePath)
  const testPaths = uniq([
    filePath,
    format({ dir, name, ext: '.json' }),
    format({ dir, name, ext: '.jsonc' }),
  ])
  const resolvedPath = testPaths.find(fileExistsSync)
  if (!resolvedPath) {
    throw new Error(
      `File path set, but none of the following tested derivations exist:\n${testPaths
        .map(p => ` - ${join(resolve('.'), p)}`)
        .join('\n')}`
    )
  }
  try {
    return JSONC.parse(readFileSync(resolvedPath, 'utf-8'))
  } catch (_error) {
    throw new Error(
      `Invalid config file in ${join(resolve('.'), resolvedPath)}`
    )
  }
}

export const createConfigStorage = (
  opts: Pick<
    ConfigLoaderOptions,
    'defaultConfigPath' | 'userConfigPath' | 'envMode'
  >
): Record<any, any> => {
  const defaultConfig = loadFile(opts.defaultConfigPath)
  const userConfig = loadFile(opts.userConfigPath)
  let envConfig: any = {}
  if (opts.envMode === 'default' || opts.envMode === 'merged') {
    const configKeys = [
      ...keys(defaultConfig),
      ...(opts.envMode === 'merged' ? keys(userConfig) : []),
    ]
    envConfig = fromPairs(
      configKeys
        .map(k => [k, process.env[k]] as [string, any])
        .filter(x => x[1] !== undefined)
    )
  } else if (opts.envMode === 'all') {
    envConfig = process.env
  }
  return { ...defaultConfig, ...userConfig, ...envConfig }
}
