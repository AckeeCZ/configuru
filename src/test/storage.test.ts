import { createConfigStorage } from 'lib/storage';
import { resolve } from 'path';

describe('storage', () => {
    test('Empty', () => {
        expect(createConfigStorage({})).toEqual({});
    });
    test('Non-existent file throws error', () => {
        expect(() => createConfigStorage({ defaultConfigPath: resolve(__dirname, './sandbox/dayum') })).toThrow(
            /exist/
        );
    });
    test('Non-valid file throws error', () => {
        expect(() => createConfigStorage({ defaultConfigPath: resolve(__dirname, './sandbox/invalid') })).toThrow(
            /valid/
        );
    });
    test('Extension not required', () => {
        const def = createConfigStorage({ defaultConfigPath: resolve(__dirname, './sandbox/default') });
        const defJson = createConfigStorage({ defaultConfigPath: resolve(__dirname, './sandbox/default.json') });
        const defJsonc = createConfigStorage({ defaultConfigPath: resolve(__dirname, './sandbox/default.jsonc') });
        expect(def).toStrictEqual(defJson);
        expect(defJson).toStrictEqual(defJsonc);
    });
    test('Default', () => {
        expect(createConfigStorage({ defaultConfigPath: resolve(__dirname, './sandbox/default.json') }))
            .toMatchInlineSnapshot(`
            Object {
              "bar": 1,
              "baz": 2,
              "defaultConfig": true,
              "foo": "default",
            }
        `);
    });
    test('Default (jsonc)', () => {
        expect(createConfigStorage({ defaultConfigPath: resolve(__dirname, './sandbox/default.jsonc') }))
            .toMatchInlineSnapshot(`
            Object {
              "bar": 1,
              "baz": 2,
              "defaultConfig": true,
              "foo": "default",
            }
        `);
    });
    test('Default & user', () => {
        expect(
            createConfigStorage({
                defaultConfigPath: resolve(__dirname, './sandbox/default.json'),
                userConfigPath: resolve(__dirname, './sandbox/user.json'),
            })
        ).toMatchInlineSnapshot(`
            Object {
              "bar": 1,
              "baz": 2,
              "defaultConfig": true,
              "foo": "user",
              "quix": false,
              "userConfig": true,
            }
        `);
    });
    test('Default & user & env', () => {
        process.env.foo = 'env';
        process.env.bar = '3';
        expect(
            createConfigStorage({
                defaultConfigPath: resolve(__dirname, './sandbox/default.json'),
                userConfigPath: resolve(__dirname, './sandbox/user.json'),
                envMode: 'default',
            })
        ).toMatchInlineSnapshot(`
                Object {
                  "bar": "3",
                  "baz": 2,
                  "defaultConfig": true,
                  "foo": "env",
                  "quix": false,
                  "userConfig": true,
                }
            `);
    });
});
