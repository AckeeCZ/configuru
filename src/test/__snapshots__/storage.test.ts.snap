// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`storage Default & user & env 1`] = `
Object {
  "bar": "3",
  "baz": undefined,
  "defaultConfig": undefined,
  "foo": "env",
  "quix": false,
  "userConfig": true,
}
`;

exports[`storage Default & user 1`] = `
Object {
  "bar": 1,
  "baz": 2,
  "defaultConfig": true,
  "foo": "user",
  "quix": false,
  "userConfig": true,
}
`;

exports[`storage Default 1`] = `
Object {
  "bar": 1,
  "baz": 2,
  "defaultConfig": true,
  "foo": "default",
}
`;
